pydicom/data/download.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2018-2020 pydicom authors. See LICENSE file for details.
+
+import functools
+import json
+import os
+import pathlib
+import urllib.request
+import warnings
+import hashlib
+import contextlib
+
+try:
+    import tqdm
+
+    class DownloadProgressBar(tqdm.tqdm):
+        def update_to(self, b=1, bsize=1, tsize=None):
+            if tsize is not None:
+                self.total = tsize
+            self.update(b * bsize - self.n)
+
+except ImportError:
+    @contextlib.contextmanager
+    def DownloadProgressBar(*args, **kwargs):
+        try:
+            class dummy:
+                def update_to(*args, **kwargs):
+                    pass
+
+            yield dummy
+        finally:
+            pass
+
+HERE = pathlib.Path(__file__).resolve().parent
+
+
+def calculate_file_hash(filename):
+    BLOCKSIZE = 65536
+    hasher = hashlib.sha256()
+    with open(filename, "rb") as f:
+        buf = f.read(BLOCKSIZE)
+        while len(buf) > 0:
+            hasher.update(buf)
+            buf = f.read(BLOCKSIZE)
+
+    return hasher.hexdigest()
+
+
+def get_config_dir():
+    config_dir = pathlib.Path.home().joinpath(".pydicom")
+    config_dir.mkdir(exist_ok=True)
+
+    return config_dir
+
+
+def download_with_progress(url, filepath):
+    with DownloadProgressBar(
+        unit="B", unit_scale=True, miniters=1, desc=url.split("/")[-1]
+    ) as t:
+        urllib.request.urlretrieve(url, filepath, reporthook=t.update_to)
+
+
+def get_data_dir():
+    data_dir = get_config_dir().joinpath("data")
+    data_dir.mkdir(exist_ok=True)
+
+    return data_dir
+
+
+@functools.lru_cache()
+def get_url_map():
+    with open(HERE.joinpath("urls.json"), "r") as f:
+        url_map = json.load(f)
```

Comment:
For the lru_cache, this means that once this module is loaded it only reads the file once. Otherwise everytime the function is loaded it will reload the file instead of using memory (this is the same as what `import filename` does, so this is more code when just `import foo` would achieve the same as this lru_cache).

For the url.json, I agree it would be equivalent to put it within a .py. The hashes file though, that's actually dynamically added to during development if a file is added to the urls file and a hash does not yet exist. To keep this functionality it would be easier to keep the hashes.json as a json file.

Let me know how you'd like me to proceed here.